### ISSUE TITLE
ENH: Create ids for CSV rows automatically

### ISSUE THREAD

**Issue by dgarijo** [oeg-upm/morph-kgc#78]
#### Is your Feature Request Related to a Problem?
When I have a CSV table without a clear identifier, I would like to have automated means for morph to create row ids without me having to manipulate the CSV

#### Describe the Solution you'd Like
A function or keyword that I could use in my mappings (e.g., {AUTO_ID}) to indicate that morph should generate ids for that row.

#### Describe Alternatives you've Considered
The current alternative is opening a csv, adding a new column with a different id for each row and saving it.

**Comment by dachafra** [comment]
@dgarijo but this feature will make your mapping dependent on a specific implementation (i.e. Morph-KGC will be the only RML engine able to parse this kind of mappings) which IMO it's not a very good idea

**Comment by dgarijo** [comment]
@dachafra, to be honest, as I user I don't care about compliance now that I see that morph-kgc works faster than the other engines I tried :)

I care about not having to do this by hand, which is what I have to do now.

While I think that your concern is valid, especially from the developer side, I feel like always going behind the spec can be detrimental to support quality-of-life features such as this one. Your suggestion to include a rule to the mapping may be a nice solution that works from both ends.

**Comment by dachafra** [comment]
Morph-KGC is the fastest now! But if in the future you may want to use another engine (e.g. you want to virtualize your data bc a new implementation is able to directly query CSV with SPARQL), so your mapping shouldn't be proprietary. 

In the end, I'm fine if @ArenasGuerreroJulian decides to implement this feature, but then, users should know that this feature is only implemented in this engine. 

In any case, I would suggest declaring a set of built-in functions for all engines when [FNML spec ](https://github.com/kg-construct/fnml-spec) will be ready

**Comment by dgarijo** [comment]
As with many other tools I get the need of relying on a standard spec. But if you add on to the standard spec for your use cases, then it should be up to others to catch up with you. I see it like having added value beyond what the standard offers, especially if you introduce quality-of-life features. But maybe this is more an offline discussion ;)

**Comment by ocorcho** [comment]
I tend to agree with @dgarijo on this aspect of adding new functionalities (and with @dachafra on the fact that this can be done using the extensbility possibility given by the spec - extensibility options are for that purpose). 
Now my question out-of-scope for morph-kgc: is it really a good practice to do auto-ids? What happens if that CSV file is updated, changes order of rows, etc., and then the URIs that you generated in the past are referring to other resources in the new re-generated RDF? Very bad design practice, which I always complain about with my students when they use the OpenRefine row identifiers for URI generation.

**Comment by ocorcho** [comment]
That would be better, since there are less options that you make a mistake on assigning tomorrow a URI that was valid today for another row (very dangerous if somebody links to you).
